src/scala/org/pantsbuild/scoverage/report/ScoverageReport.scala

```diff
@@ -0,0 +1,254 @@
+
+  // Copying the logger from org.pantsbuild.zinc.bootstrap
+  // As per https://github.com/pantsbuild/pants/issues/6160, this is a workaround
+  // so we can run zinc without $PATH (as needed in remoting).
+  System.setProperty("sbt.log.format", "true")
+
+  // Setting the logger
+  val logger = ConsoleLogger.apply()
+
+  case class ReportOptions(
+    loadDataDir: Boolean,
+    measurementsPath: String,
+    sourcePath: String,
+    reportPath: String,
+    dataPath: String,
+    writeHtml: Boolean,
+    writeXml: Boolean,
+    writeXmlDebug: Boolean,
+    cleanOld: Boolean,
+    targetFilters: Seq[String]) {
+    val writeReports: Boolean = {
+      val write = writeHtml || writeXml
+      if (!write) {
+        // we could end here, but we will still attempt to load the coverage and dump stats as info
+        // to the log, so maybe somebody would want to run this w/o generating reports?
+        logger.warn("No report output format specified, so no reports will be written.")
+      }
+      write
+    }
+  }
+
+  object ReportOptions {
+    def apply(settings: Settings): ReportOptions = {
+      ReportOptions(
+        loadDataDir = settings.loadDataDir,
+        measurementsPath = settings.measurementsDirPath,
+        sourcePath = settings.sourceDirPath,
+        reportPath = settings.reportDirPath,
+        dataPath = settings.dataDirPath,
+        writeHtml = settings.writeHtmlReport,
+        writeXml = settings.writeXmlReport,
+        writeXmlDebug = settings.writeXmlDebug,
+        cleanOld = settings.cleanOldReports,
+        targetFilters = settings.targetFilters)
+    }
+  }
+  /**
+   *
+   * @param dataDirs list of measurement directories for which coverage
+   *                 report has to be generated
+   * @return         [Coverage] object for the [dataDirs]
+   */
+  private def aggregatedCoverage(dataDirs: Seq[File]): Coverage = {
+    var id = 0
+    val coverage = Coverage()
+    dataDirs foreach { dataDir =>
+      val coverageFile: File = Serializer.coverageFile(dataDir)
+      if (coverageFile.exists) {
+        val subcoverage: Coverage = Serializer.deserialize(coverageFile)
+        val measurementFiles: Array[File] = IOUtils.findMeasurementFiles(dataDir)
+        val measurements = IOUtils.invoked(measurementFiles.toIndexedSeq)
+        subcoverage.apply(measurements)
+        subcoverage.statements foreach { stmt =>
+          // need to ensure all the ids are unique otherwise the coverage object will have stmt collisions
+          id = id + 1
+          coverage add stmt.copy(id = id)
+        }
+      }
+    }
+    coverage
+  }
+
+  /**
+   *
+   * @param dataDirs list of measurement directories for which coverage
+   *                 report has to be generated
+   * @return         Coverage object wrapped in Option
+   */
+  private def aggregate(dataDirs: Seq[File]): Option[Coverage] = {
+    logger.info(s"Found ${dataDirs.size} subproject scoverage data directories [${dataDirs.mkString(",")}]")
```

Comment:
If dataDirs is null, this will throw an NullPointerException

Reply:
Feels like this method could just be consolidated with aggregatedCoverage

Reply:
`aggregate` is called only by `loadAggregatedCoverage` and I am checking there if the `dataDirs` is null or not. If it doesn't exist, it raises a runtime exception.